hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java

```diff
@@ -105,15 +101,15 @@ private void processPipelineReport(PipelineReport report, DatanodeDetails dn)
     }
 
     if (pipeline.getPipelineState() == Pipeline.PipelineState.ALLOCATED) {
-      LOGGER.info("Pipeline {} reported by {}", pipeline.getId(), dn);
-      pipeline.reportDatanode(dn);
-      if (pipeline.isHealthy()) {
-        // if all the dns have reported, pipeline can be moved to OPEN state
+      LOGGER.info("Pipeline {} reported by {} isLeader?: {}",
+          pipeline.getId(), dn, report.getIsLeader());
+
+      if (report.getIsLeader()) {
+        // Pipeline reported as the leader
+        pipeline.setLeaderId(dn.getUuid());
```

Comment:
We can update the `LeaderId` even if the pipeline is in OPEN state. There is a possibility that the leader might have changed.

The whole logic can be simplified like below:
```
pipeline.reportDatanode(dn);
if (report.getIsLeader()) {
  pipeline.setLeaderId(dn.getUuid());
}
if ((pipeline.getPipelineState() == Pipeline.PipelineState.ALLOCATED) && pipeline.isHealthy()) {
  pipelineManager.openPipeline(pipelineID);
}
```

Reply:
Good catch, the bug got introduced during last set of changes.